<!doctype html><html lang="en"><head><meta charset="UTF-8"><title>ES-NDD Documentation</title><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4.6.0/dist/css/bootstrap.min.css" integrity="sha384-B0vP5xmATw1+K9KRQjQERJvTumQW0nPEzvF6L/Z6nronJ3oUOFUFpCjEUQouq2+l" crossorigin="anonymous"><link rel="stylesheet" href="css/394c0765f740a217d3e07512a7cd850b.css"><script defer="defer" src="bundle.js"></script></head><body><nav class="navbar navbar-dark static-top navbar-expand-lg"><a class="navbar-brand" href="../index.html">ES-NDD</a> <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation"><span class="navbar-toggler-icon"></span></button><div class="navbar-collapse collapse w-100 order-3 dual-collapse2" id="navbarNav"><ul class="navbar-nav ml-auto"><li class="nav-item"><a class="nav-link" href="documentation.html">Documentation</a></li><li class="nav-item"><a class="nav-link" href="contact.html">Contact</a></li></ul></div></nav><header class="masthead text-dark text-center"><div class="container"><div class="row"><div class="col-xl-9 mx-auto"><h2 class="text-center font-weight-normal">Documentation</h2></div></div></div></header><div class="container"><div class="row"><div class="col-xl-9 mx-auto"><h2>Summary</h2><p>Neurodevelopmental disorders (NDDs) are a heterogeneous group of conditions involving various forms of disruption to brain development. Over hundreds of genes have been found to be associated with NDDs thus far. However, NDDs are complex and mutations in NDD associated genes are not straightforward to conceptualize at the molecular level, for which three-dimensional (3D) structures of proteins provide an informative mean.</p><p>Here, we present the ES-NDD. The tool provides access to the aggregation of ~1,200 experimentally and computationally modeled 3D monomeric structures as well as multimeric protein complexes and their characterization using a consensus approach to identity Essential3D sites are conserved across human paralogs. With ES-NDD a user is able to interactively explore Essential3D sites, that are paralog conserved, population constraint and enriched for pathogenic variants in the 3D structures, in the context of 41,939 population variants aggregated from gnomAD, 6,343 pathogenic (likely-pathogenic) variants from ClinVar and HGMD as well as functional and domain annotations from Uniprot.</p><p>For details please refer to the associated <a href="#">publication</a>.</p><h2 class="mt-4">How to use the Essential3D Viewer</h2><img src="img/docs1.png" class="img-fluid m-5"><ol><li>A) A user selects the gene of interest, (<span class="docs-letter">B</span>) selects one of the available structure types and (<span class="docs-letter">C</span>) decides which protein chain (for multimeric-PDB set) of the protein structure will be chosen.</li><img src="img/docs2.png" class="img-fluid m-5"><li>Now a user can choose any available annotation from the bar (<span class="docs-letter">A</span>), to display single 3D-scores as well as our consensus Essential3D annotation to discover essential regions on the linear protein sequence (<span class="docs-letter">B</span>) and on the selected protein structure (<span class="docs-letter">C</span>). Further, to bring the highlighted regions into context of available variants, variants from gnomAD, ClinVar and HGMD can be visualized in the 3D protein structure together with several functional and domain annotations from Uniprot.</li><li>To access all annotations on the protein structure the corresponding Pymol session (<span class="docs-letter">D</span>) can be downloaded for manual inspection. Alternatively, the annotations for the gene can be downloaded as table (<span class="docs-letter">E</span>). Complete access to all Essential3D annotations as well as individual 3D-scores for all genes and datasets are available in github .</li></ol></div></div></div></body></html>